sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala

```diff
@@ -19,274 +19,575 @@ package org.apache.spark.sql.execution.streaming
-    iterator.foreach { kv =>
-      store.remove(kv.key, ttlColumnFamilyName)
+      store.remove(elementKey, MIN_INDEX)
+      deleteFromTTLIndex(existingMinExpiration, elementKey)
     }
   }
 
-  def upsertTTLForStateKey(
-      expirationMs: Long,
-      groupingKey: UnsafeRow,
-      userKey: UnsafeRow): Unit = {
-    val encodedTtlKey = keyRowEncoder.encodeTTLRow(
-      expirationMs, groupingKey, userKey)
-    store.put(encodedTtlKey, EMPTY_ROW, ttlColumnFamilyName)
+  // Exposed for testing.
+  private[sql] def minIndexIterator(): Iterator[(UnsafeRow, Long)] = {
+    store
+      .iterator(MIN_INDEX)
+      .map(kv => (kv.key, kv.value.getLong(0)))
   }
+}
+
+
+/**
+ * OneToOneTTLState is an implementation of [[TTLState]] that is used to manage
+ * TTL for state variables that need a single secondary index to efficiently manage
+ * records with an expiration.
+ *
+ * The primary index for state variables that can use a [[OneToOneTTLState]] have
+ * the form of: [elementKey -> (value, elementExpiration)]. You'll notice that, given
+ * a timestamp, it would take linear time to probe the primary index for all of its
+ * expired values.
+ *
+ * As a result, this class uses helper methods from [[TTLState]] to maintain the secondary
+ * index from [(elementExpiration, elementKey) -> EMPTY_ROW].
+ *
+ * For an explanation of why this structure is not always sufficient (e.g. why the class
+ * [[OneToManyTTLState]] is needed), please visit its class-doc comment.
+ */
+abstract class OneToOneTTLState(
+    stateNameArg: String,
+    storeArg: StateStore,
+    elementKeySchemaArg: StructType,
+    ttlConfigArg: TTLConfig,
+    batchTimestampMsArg: Long,
+    metricsArg: Map[String, SQLMetric]) extends TTLState {
+  override def stateName: String = stateNameArg
+  override def store: StateStore = storeArg
+  override def elementKeySchema: StructType = elementKeySchemaArg
+  override def ttlConfig: TTLConfig = ttlConfigArg
+  override def batchTimestampMs: Long = batchTimestampMsArg
+  override def metrics: Map[String, SQLMetric] = metricsArg
 
   /**
-   * Clears any state which has ttl older than [[ttlExpirationMs]].
+   * This method updates the TTL for the given elementKey to be expirationMs,
+   * updating both the primary and secondary indices if needed.
+   *
+   * Note that an elementKey may be the state variable's grouping key, _or_ it
+   * could be a composite key. MapState is an example of a state variable that
+   * has composite keys, which has the structure of the groupingKey followed by
+   * the specific key in the map. This method doesn't need to know what type of
+   * key is being used, though, since in either case, it's just an UnsafeRow.
+   *
+   * @param elementKey the key for which the TTL should be updated, which may
+   *                   either be an UnsafeRow derived from [[SingleKeyTTLRow]]
+   *                   or [[CompositeKeyTTLRow]].
+   * @param expirationMs the new expiration timestamp to use for elementKey.
```

Comment:
Yes, good point. And also, the comment for `elementKey` was wrong, since I removed `CompositeKeyTTLRow`.

> This seems to give another aspect of this base class

Yes, that was an intentional design decision that I made:

```
 * [2]. You might also ask, why design it this way? We want the TTLState abstract
 *      sub-classes to write to both the primary and secondary indexes, since they
 *      both need to stay in sync; co-locating the logic is cleanest.
```

Reply:
But this gives away the goodness of mix-in and require the interface to be coupled with the usage. I'm not intended to say which one is better, but worth thinking of trade-off.

Reply:
Very fair comment. It's like a trait, since it provides utility methods to write to the TTL index/work queue. But it also isn't super mixing-in, in that it also writes to the primary index.

So the trade-off is:

- Couple the logic, break the pattern of mix-in.
- Decouple the logic, but have a "clean" mix-in.

I do like coupling the logic, since it is more likely to be correct and not have a consistency issue. If you feel strongly, let's continue the discussion on this.

Reply:
Either is fine for me as it's really a trade off and I guess the change would be non-trivial to switch.

Reply:
Ok, I will keep the existing behavior of having the trait manage both indices. I've made the rename to the method. If we want to do a refactor later (that doesn't change the format), then we can.